### ISSUE TITLE
Using MRR  as the evaluation metric for TGN

### ISSUE THREAD

**Issue by ZeroerWiser** [pyg-team/pytorch_geometric#7558]
### 🚀 The feature, motivation and pitch

May I ask how I can use MRR as the evaluation index to get the performance evaluation of TGN model [https://github.com/pyg-team/pytorch_geometric/blob/master/examples/tgn.py](url) Thank you very much. In the code I have used so far, it seems that using a batch (e.g., 200) of candidate destination nodes in the MRR calculation results in a final MRR value close to 1, which is obviously not possible. I sincerely look forward to your reply. It would be better if there is a reference code.

**Comment by rusty1s** [comment]
I think this would be straightforward to add. We would need to sample multiple negatives per source node, rank them, and then compute the MRR on top of this. Let me know if you already have an implementation of this. Happy to take a look.

**Comment by ZeroerWiser** [comment]
> I think this would be straightforward to add. We would need to sample multiple negatives per source node, rank them, and then compute the MRR on top of this. Let me know if you already have an implementation of this. Happy to take a look.

well, the result of this MRR is 1 for the code running, so I suspect this is caused by processing mrr candidate list inside the batch (i.e., the destination sequence size is the size of the batch, and I need to calculate the likelihood that the source node is connected to all nodes except itself).my code is under following:
torch.manual_seed(1234)
@torch.no_grad()
def test(loader, topk=5):
memory.eval()
gnn.eval()
link_pred.eval()
torch.manual_seed(12345)  # Ensure deterministic sampling across epochs.
mrrs = []
for batch in loader:
batch = batch.to(device)
src, pos_dst, t, msg = batch. src, batch. dst, batch. t, batch.msg
print('src:',src)
print('src_len',len(src))
neg_dst = torch. randint(min_dst_idx, max_dst_idx + 1, (src.size(0), ),
dtype=torch. long, device=device)
n_id = torch. cat([src, pos_dst, neg_dst]).unique()
n_id, edge_index, e_id = neighbor_loader(n_id)
assoc[n_id] = torch.arange(n_id.size(0), device=device)
print('assoc:',assoc)
print('assoc_lem:',len(assoc))
z, last_update = memory(n_id)
z = gnn(z, last_update, edge_index, events.t[e_id].to(device),
events.msg[e_id].to(device))

pos_out = link_pred(z[assoc[src]], z[assoc[pos_dst]])
neg_out = link_pred(z[assoc[src]], z[assoc[neg_dst]])

y_pred = torch. cat([pos_out, neg_out], dim=0).sigmoid().cpu()
print('pred:',y_pred)
print('len:',y_pred.shape)
y_true = torch.cat(
[torch.ones(pos_out.size(0)),
torch.zeros(neg_out.size(0))], dim=0)
#print('true:',y_true)
pos_indices = (y_true == 1).nonzero(as_tuple=False)[:, 0]
target_probs = y_pred[pos_indices]
sorted_indices = y_pred.view(-1).argsort(descending=True).numpy()
#print(sorted_indices)
topk_indices = sorted_indices[:topk]
print(topk_indices[0],topk_indices)
#print(topk_indices)
mrrs.append((1.0 / (torch. where(target_probs >  0)[0][0].item() + 1)) if target_probs.size(0) != 0\
and torch.squeeze(target_probs) in y_pred[topk_indices] else 0.0)
memory. update_state(src, pos_dst, t, msg)
neighbor_loader. insert(src, pos_dst)
return float(torch.tensor(mrrs).mean())

**Comment by rusty1s** [comment]
What happens if you increase the number of negatives you sample per each positive?

**Comment by rusty1s** [comment]
Sorry for late reply. I still think you just have to increase the number of negatives to sample, e.g.,
```python
neg_dst = torch. randint(min_dst_idx, max_dst_idx + 1, (5 * src.size(0), ),
```
Then, you can perform link prediction for a set of negatives
```python
neg_out = link_pred(torch.cat([z[assoc[src]] * 5]), z[assoc[neg_dst]])
```
and rank the positives across the negatives accordingly afterwards.
